### ISSUE TITLE
[Solved] Problems about Load Balance

### ISSUE THREAD

**Issue by wakeupnuts** [princeton-sns/serval#1]
I had four PCs to build the system as the QuickGuide suggested, as follows
...............................Server1 (10.21.5.50)
.............................. /
 Client ------------ LB (10.21.5.26)
(10.27.0.139).......... \
............................... Server2 (10.21.5.51)

Where LB, Server1 and Server2 are in the same IP broadcast domain, all connected to a TP-Link switcher(TL-SF1008+), while Client is in other IP broadcast domain. All the PCs run Ubuntu 12.04, except that LB run Ubuntu 11.10.

I set up the servers, each server has an instance of tcp_server and are waiting to transfer a file about 1.6G. Then I customize LB by executing the following commands on LB as the Quickstart-Guide instructed
$ ./src/tools/serv service add 76568 10.21.5.50
$ ./src/tools/serv service add 76568 10.21.5.51
/\* I change to root then chmod the files */
$ echo 1 > /proc/sys/net/ipv4/ip_forward
$ echo 1 > /proc/sys/net/serval/sal_forward

Then I launch the tcp_client on client and ask for the file transfer service. The client is not in the IP broadcast domain where LB and servers are in, so I add LB's IP address to Client's service routing rules first, as follows
$ ./src/tools/serv service add 76568 10.21.5.26 (the IP address of LB)
$ ./src/test/tcp_client -s 76568 -f ~/tmp/receive_file00

then immediately re-ask the service by execute the command again
$ ./src/test/tcp_client -s 76568 -f ~/tmp/receive_file01

The 1.6G file takes about 3-5 minutes to finish transmission, so when I retype the tcp_client command seconds after the first time, the first service session is not end yet, and IT'S EXPECTED that the second instance of tcp_client should be served by another server (idle one). HOWEVER, it didn't happen. The 2nd  tcp_client instance simply get response from the server who serve the first instance(In this case, server 10.21.5.50) and wait. No LoadBalance effects showed.

---

Then I thought about the experiment. To my great surprise, it seems the experiment has some LOGICAL FLAW ----- The dedicated load balancer adds routing rules on itself, and then deals with only the first packets. What does it mean? It means the load balancer WILL NOT KNOW THE LOAD at all, then how could the LB perform load balancing without the knowledge of the loads? 

Do I miss something or make something wrong in the experiment? Why the experiment of Load Balance is seemingly to fail inevitably?

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++

see my last comment for the solution.

**Comment by erimatnor** [comment]
The current load balancing scheme we implement for service routers is using weighted proportional split. This means that a server is randomly chosen with some likelihood proportional to the weight, and it is entirely possible that two subsequent requests hit the same server. It might take a number of subsequent requests to see a balancing effect. Maybe this is what you are experiencing?

I am not sure what you mean by "dedicated load balancer adds routing rules on itself" and "the load balancer WILL NOT KNOW THE LOAD". Yes, it is true that the LB only touches the first packet, so it won't know the number of ongoing flows to each server. But this is by design, and is enough to balance the requests across the pool of servers. Knowing the actual load on the servers would require monitoring schemes that feed information to the LB, but that is out of scope of Serval.

**Comment by wakeupnuts** [comment]
Thank you very much for your reply!

I supposed that Serval would monitor the servers' load and expected two subsequent requests would hit different servers. In my experiment, I only launched 2-3 consecutive requests each time. I've repeated the load-balance experiment 4 times and found all requests goes to the same server every time. Indeed, as you said, it's possible that these subsequent requests hit the same server while the scheme worked properly. 

I will re-do the experiment to test Serval's load-balancing function. I do believe that the size of file used for transferring is irrelevant to the final results, since LB won't monitor the server's load. 

Your response really helps, thanks a lot!

**Comment by wakeupnuts** [comment]
After knowing that Serval's load-balancing function is based on weighted proportional random selection, I realized that my original assumption that Serval's load-balancer(LB) would supervise the load on each server is wrong, and I need to redesign my experiment to test the load balancing functionality. Here is what I do:

I decide that the service requests must be repeated numerous times, a dozen of consecutive requests are far less than enough to observe the effect of load-balancing, I planned to repeat the requests for 100, 1000 and 5000 times.

In order to efficiently generate the requests and to collect data, I trimmed the tcp_server/tcp_client of the test directory into a much simpler pair date_server/date_client, the date_client simply reads the time in date_server and then shutdown, with an output suitable for data collection. (I put the codes in https://github.com/wakeupnuts/serval)

The result justify Serval's load balancing function. Here's the result of 100 repetitions: (10.21.5.50 & 10.21.5.51 are the IP addresses of the two servers)
10.21.5.50 : 1
10.21.5.51 : 2
10.21.5.50 : 3
10.21.5.51 : 4
10.21.5.51 : 5
10.21.5.50 : 6
10.21.5.50 : 7
10.21.5.50 : 8
10.21.5.50 : 9
10.21.5.50 : 10
10.21.5.51 : 11
10.21.5.51 : 12
10.21.5.51 : 13
10.21.5.51 : 14
10.21.5.50 : 15
10.21.5.51 : 16
10.21.5.50 : 17
10.21.5.51 : 18
10.21.5.50 : 19
10.21.5.50 : 20
10.21.5.51 : 21
10.21.5.51 : 22
10.21.5.50 : 23
10.21.5.50 : 24
10.21.5.51 : 25
10.21.5.51 : 26
10.21.5.51 : 27
10.21.5.51 : 28
10.21.5.51 : 29
10.21.5.51 : 30
10.21.5.51 : 31
10.21.5.50 : 32
10.21.5.50 : 33
10.21.5.50 : 34
10.21.5.51 : 35
10.21.5.50 : 36
10.21.5.50 : 37
10.21.5.51 : 38
10.21.5.51 : 39
10.21.5.50 : 40
10.21.5.51 : 41
10.21.5.50 : 42
10.21.5.51 : 43
10.21.5.51 : 44
10.21.5.50 : 45
10.21.5.51 : 46
10.21.5.50 : 47
10.21.5.50 : 48
10.21.5.51 : 49
10.21.5.50 : 50
10.21.5.50 : 51
10.21.5.51 : 52
10.21.5.51 : 53
10.21.5.50 : 54
10.21.5.51 : 55
10.21.5.51 : 56
10.21.5.51 : 57
10.21.5.51 : 58
10.21.5.50 : 59
10.21.5.51 : 60
10.21.5.51 : 61
10.21.5.51 : 62
10.21.5.51 : 63
10.21.5.50 : 64
10.21.5.50 : 65
10.21.5.50 : 66
10.21.5.50 : 67
10.21.5.50 : 68
10.21.5.51 : 69
10.21.5.51 : 70
10.21.5.51 : 71
10.21.5.50 : 72
10.21.5.51 : 73
10.21.5.51 : 74
10.21.5.51 : 75
10.21.5.51 : 76
10.21.5.50 : 77
10.21.5.51 : 78
10.21.5.51 : 79
10.21.5.51 : 80
10.21.5.51 : 81
10.21.5.51 : 82
10.21.5.50 : 83
10.21.5.50 : 84
10.21.5.50 : 85
10.21.5.50 : 86
10.21.5.50 : 87
10.21.5.50 : 88
10.21.5.50 : 89
10.21.5.51 : 90
10.21.5.51 : 91
10.21.5.50 : 92
10.21.5.50 : 93
10.21.5.50 : 94
10.21.5.51 : 95
10.21.5.51 : 96
10.21.5.51 : 97
10.21.5.51 : 98
10.21.5.50 : 99
10.21.5.50 : 100
